twittermap/public/javascripts/cache/services.js

```diff
@@ -0,0 +1,321 @@
+ it gets the requested region data from the middleware.*/
+  this.getCityPolygonsFromCache = function city(bounds){
+
+       var deferred = new $.Deferred();
+       var data_response;
+
+
+       var rteBounds = "city/" + bounds._northEast.lat + "/" + bounds._southWest.lat + "/" + bounds._northEast.lng + "/" + bounds._southWest.lng;
+
+       var currentRequestPolygon = turf.polygon([[
+                                    [bounds._northEast.lng,bounds._northEast.lat],
+                                    [bounds._northEast.lng,bounds._southWest.lat],
+                                    [bounds._southWest.lng,bounds._southWest.lat],
+                                    [bounds._southWest.lng,bounds._northEast.lat],
+                                    [bounds._northEast.lng,bounds._northEast.lat]
+                                ]]);
+
+       var bbox = turf.bbox(currentRequestPolygon);
+        // to search in Rbush Tree ,we need the MBR of the requested region.
+       var item = {
+           minX : bbox[0],
+           minY : bbox[1],
+           maxX : bbox[2],
+           maxY : bbox[3]
+       }
+
+
+       if(typeof cachedRegion != "undefined" && typeof turf.difference(currentRequestPolygon,cachedRegion) == "undefined") {
+              //cache HIT
+
+             var result = cachedCityPolygonTree.search(item);
+             data_response = turf.featureCollection(result);
+             deferred.resolve(data_response);
+
+             return deferred.promise();
+
+       }else{
+                   //cache MISS
+
+              var buffered = turf.buffer(currentRequestPolygon, preFetchDistance , 'miles');
+              var bboxBuffer = turf.bbox(buffered);
+                   //Pre Fetch
+              var rteExtends = "city/" + bboxBuffer[3] + "/" + bboxBuffer[1] + "/" + bboxBuffer[2] + "/" + bboxBuffer[0];
+
+              $http.get(rteExtends).success(function(data) {
+
+                               var result_set = insertIntoTree(data.features,currentRequestPolygon).done(function(){
+
+                                                var extendPoly = turf.bboxPolygon(bboxBuffer);
+
+                                                if(cachedRegion == undefined)
+                                                    {cachedRegion = extendPoly;}
+                                                else
+                                                    {cachedRegion = turf.union(extendPoly,cachedRegion);}
+
+                                                var result = cachedCityPolygonTree.search(item);
+                                                data_response = turf.featureCollection(result);
+
+                                                deferred.resolve(data_response);
+
+
+                      });
+              }).error(function(data) {
+                                    console.error("Load city data failure");
```

Comment:
This was something that was present in earlier codebase whenever They were sending HTTP requests ,They have this error function if middleware had failed.